org.eclipse.jdt.ls.debug/src/org/eclipse/jdt/ls/debug/adapter/ProtocolServer.java

```diff
@@ -0,0 +1,229 @@
+        message.seq = this.sequenceNumber++;
+
+        String jsonMessage = JsonUtils.toJson(message);
+        char[] jsonBytes = jsonMessage.toCharArray();
+
+        String header = String.format("Content-Length: %d%s", jsonBytes.length, TWO_CRLF);
+        char[] headerBytes = header.toCharArray();
+
+        char[] data = new char[headerBytes.length + jsonBytes.length];
+        System.arraycopy(headerBytes, 0, data, 0, headerBytes.length);
+        System.arraycopy(jsonBytes, 0, data, headerBytes.length, jsonBytes.length);
+
+        try {
+            Logger.logInfo("\n[[RESPONSE]]\n" + new String(data));
+            this.writer.write(data, 0, data.length);
+            this.writer.flush();
+        } catch (IOException e) {
+            Logger.logException("Write data to io exception", e);
+        }
+    }
+
+    private void processData() {
+        while (true) {
+            if (this.bodyLength >= 0) {
+                if (this.rawData.length() >= this.bodyLength) {
+                    char[] buf = this.rawData.removeFirst(this.bodyLength);
+                    this.bodyLength = -1;
+                    dispatchRequest(new String(buf));
+                    continue;
+                }
+            } else {
+                String body = this.rawData.getString();
+                int idx = body.indexOf(TWO_CRLF);
+                if (idx != -1) {
+                    Matcher matcher = CONTENT_LENGTH_MATCHER.matcher(body);
+                    if (matcher.find()) {
+                        this.bodyLength = Integer.parseInt(matcher.group(1));
+                        this.rawData.removeFirst(idx + TWO_CRLF.length());
+                        continue;
+                    }
+                }
+            }
+            break;
+        }
+    }
+
+    private void dispatchRequest(String request) {
+        try {
+            Logger.logInfo("\n[REQUEST]\n" + request);
+            Messages.Request message = JsonUtils.fromJson(request, Messages.Request.class);
+            if (message.type.equals("request")) {
+                synchronized (this.lock) {
+                    this.isDispatchingData = true;
+                }
+
+                try {
+                    Messages.Response response = this.debugAdapter.dispatchRequest(message);
+                    if (message.command.equals("disconnect")) {
+                        this.stop();
+                        this.debugAdapter = null;
+                    }
+                    sendMessage(response);
+                } catch (Exception e) {
+                    Logger.logException("Dispatch debug protocol error", e);
+                }
+            }
+        } finally {
+            synchronized (this.lock) {
+                this.isDispatchingData = false;
+            }
+
+            while (this.eventQueue.peek() != null) {
+                sendMessage(this.eventQueue.poll());
+            }
+        }
+    }
+
+    class CharBuffer {
```

Comment:
Of course we could use java.nio.CharBuffer to implement similar effects. But java.nio.CharBuffer is more like a char array with fixed capacity.  You could put and get char sequence on it.

And our customized CharBuffer is like a linked char list, you could removeFirst and append infinitely. Here in our scenario, the customized CharBuffer is more matched.